### PR TITLE
Adding Submodule

### PR DESCRIPTION
**Pull request by stevenjoezhang**
# Breaking Change:
Install new modules:
```
git submodule init fancybox pangu
git submodule update
```

## PR Checklist
**Please check if your PR fulfills the following requirements:**

- [x] The commit message follows [our guidelines](https://github.com/theme-next/hexo-theme-next/blob/master/.github/CONTRIBUTING.md).
- [x] Tests for the changes was maked (for bug fixes / features).
   - [x] Muse | Mist have been tested.
   - [x] Pisces | Gemini have been tested.
- [ ] Docs in [NexT website](https://theme-next.org/docs/) have been added / updated (for new features).

## PR Type
**What kind of change does this PR introduce?**

- [ ] Bugfix.
- [x] Feature.
- [ ] Code style update (formatting, local variables).
- [ ] Refactoring (no functional changes, no api changes).
- [ ] Build related changes.
- [ ] CI related changes.
- [ ] Documentation content changes.
- [ ] Other... Please describe:

## What is the current behavior?

Issue resolved: #413

## What is the new behavior?

- Screenshots with this changes: N/A
- Link to demo site with this changes: N/A

### How to use?
In NexT `_config.yml`:
```yml
...
```

## Does this PR introduce a breaking change?
- [x] Yes.
- [ ] No.

### CODE REVIEW

**Comment by 1v9** [comment]
fancybox3 & gitment?

File: .gitmodules [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+	url = https://github.com/theme-next/theme-next-fancybox
+[submodule "pace"]
+	path = source/lib/pace
+	url = https://github.com/theme-next/theme-next-pace
+[submodule "jquery_lazyload"]
+	path = source/lib/jquery_lazyload
+	url = https://github.com/theme-next/theme-next-jquery-lazyload
+[submodule "fastclick"]
+	path = source/lib/fastclick
+	url = https://github.com/theme-next/theme-next-fastclick3
```
Comment (1v9) [inline .gitmodules]: fastclick3

**Comment by stevenjoezhang** [comment]
Gitment is not designed to be cloned. It just provides a jsDelivr CDN, because the origin Gitment repo is not maintained
fancybox3 and fancybox both uses /source/lib/fancybox, to avoid potential conflict, I choose fancybox3

**Comment by 1v9** [comment]
Need update much docs README...

**Comment by 1v9** [comment]
https://github.com/theme-next/theme-next-fancybox3/tree/d8a8c05db4a9426a4f56c7543da5c6efb43843de

**Comment by stevenjoezhang** [comment]
Seems tarvis-ci does not know how to clone a repo with ssl

**Comment by ivan-nginx** [comment]
Don't totally understand: for what this needed?

**Comment by JiangTJ** [comment]
> Don't totally understand: for what this needed?

Confusion too. Submodule can be done by user.

**Comment by stevenjoezhang** [comment]
This allows users installing dependencies using a simple command, no longer need to check and follow the instructions in README.md of these ‘theme-next-*’ repos

**Comment by JiangTJ** [comment]
> This allows users installing dependencies using a simple command, no longer need to check and follow the instructions in README.md of these ‘theme-next-*’ repos

Some GUI tools would clone all submodule when once clone, such as GitHub desktop......
But I don't use all.

**Comment by LEAFERx** [comment]
Personally, instead of git submodule, I prefer that plugins be hexo plugins (npm packages), and insert to theme using hexo tags.
For example, if you want jquery, you can create a plugin that provides {% jquery-script %} called hexo-jquery and users can just simply install it and change _config.yml to set useJquery: Ture.
It is convenient for users and can be simply reused by other themes.
Think about that lol. Why a theme plugin is not a hexo plugin?

**Comment by LEAFERx** [comment]
check this out for a little example:
https://petrstepanov.wordpress.com/2018/11/07/custom-hexo-tag-plugin-with-a-nunjucks-template/

**Comment by ivan-nginx** [comment]
Guys, check [here](https://github.com/theme-next/hexo-theme-next/issues/655#issuecomment-470414450) before resume discussion of this PR.

**Comment by LEAFERx** [comment]
> Guys, check [here](https://github.com/theme-next/hexo-theme-next/issues/655#issuecomment-470414450) before resume discussion of this PR.

Yes it's a great view. And automatic download may be easier to be done by npm packages lol.

**Comment by 1v9** [comment]
Philosophical debate is coming🤔. I think if there's not a good solution just don't change anything.

**Comment by ivan-nginx** [comment]
> automatic download may be easier to be done by npm packages

There is some problems with it:

1. `node_modules` path is different from `themes` path. But it's not critical and may be solved, of course.
2. For now NexT used Netlify for deploy and each time during generate Netlify reinstall `node_modules`, but not NexT theme. See logs [here](https://app.netlify.com/sites/theme-next/deploys/5c80b81e57f6ce000988c5e2) and [here](https://app.netlify.com/sites/theme-next/deploys/5c80aa131758240008537acf), for example. In 1st logs NexT just updated needed files; in 2nd – don't touch because «Already up to date». I'm talking about traffic, Netlify give 100GB per month.

But maybe `node_modules` cached somehow, I'm not sure. However, as an option we can try to move modules from Git repos to NPM.

**Comment by JiangTJ** [comment]
About npm. 
I think the main problem is `swig`. Something written in `swig` file. How to solved it?

**Comment by ivan-nginx** [comment]
What `swig`? NPM can include any files. Moreover, almost all functions can be moved from `swig` to `js` files in `scripts` directory. But on it will wasted some time, of course.

**Comment by JiangTJ** [comment]
Such as comments:
```swig
{% if theme.disqus.enable %}
  {% include 'disqus.swig' %}
{% elif theme.changyan.enable and theme.changyan.appid and theme.changyan.appkey %}
......
```
We need config in comment [index.swig](https://github.com/theme-next/hexo-theme-next/blob/master/layout/_third-party/comments/index.swig), If we use npm, need wasted some time to add code in the right place. But it's not easy.

**Comment by JiangTJ** [comment]
> 100GB per month

Only when a user visits a page will it consume. So don't worry reinstall `node_modules`.

**Comment by ivan-nginx** [comment]
> We need config in comment index.swig, If we use npm, need wasted some time to add code in the right place.

Maybe. But u talking now about NexT. Need to make decision about modules firstly.

**Comment by LEAFERx** [comment]
By hexo plugins I mean install in root of the blog, not in NexT.
I'm working on the leancloud conuter refactor including modulize and automatic setup. I can show you how I design the module to work when I'm finished. Hope I can finish before next week.

> Such as comments:
> 
> ```
> {% if theme.disqus.enable %}
>   {% include 'disqus.swig' %}
> {% elif theme.changyan.enable and theme.changyan.appid and theme.changyan.appkey %}
> ......
> ```
> We need config in comment [index.swig](https://github.com/theme-next/hexo-theme-next/blob/master/layout/_third-party/comments/index.swig), If we use npm, need wasted some time to add code in the right place. But it's not easy.

Sure current code need refactor. But we can do it step by step and still keep those small pieces inside. And automatic install dependencies for NexT does excite me a lot.

> > automatic download may be easier to be done by npm packages
> 
> There is some problems with it:
> 
> 1. `node_modules` path is different from `themes` path. But it's not critical and may be solved, of course.
> 2. For now NexT used Netlify for deploy and each time during generate Netlify reinstall `node_modules`, but not NexT theme. See logs [here](https://app.netlify.com/sites/theme-next/deploys/5c80b81e57f6ce000988c5e2) and [here](https://app.netlify.com/sites/theme-next/deploys/5c80aa131758240008537acf), for example. In 1st logs NexT just updated needed files; in 2nd – don't touch because «Already up to date». I'm talking about traffic, Netlify give 100GB per month.
> 
> But maybe `node_modules` cached somehow, I'm not sure. However, as an option we can try to move modules from Git repos to NPM and check it for traffic.
> 
> BTW, if we talking about NPM, maybe NexT can live on NPM too? But it's a little bit extra work: make release on Git, then make release on NPM. Need to think about it.

Netlify do cache npm modules. see [here](https://www.netlify.com/docs/build-gotchas/#build-cache)

**Comment by stevenjoezhang** [comment]
Superseded by https://github.com/next-theme/plugins

**Pull request closed without merging**